[SRVKS-1306] Configuring long running requests (#89935)
openshift/openshift-docs: modules/configuring-default-route-timeouts-globally.adoc

Reviewer: For these longer commands, I would suggest breaking them across lines so more of the command is easy to see without scrolling

// Module included in the following assemblies:
//
// * knative-serving/config-applications/configuring-revision-timeouts.adoc
:_content-type: PROCEDURE
[id="configuring-default-route-timeouts-globally_{context}"]
= Configuring the default route timeouts globally

By configuring the route timeouts globally, you can ensure consistent timeout settings across all services, simplifying management for workloads that have similar timeout needs and reducing the need for individual adjustments.

.Procedure

You can configure the route timeouts globally by updating the `ROUTE_HAPROXY_TIMEOUT` environment value in your `serverless-operator` subscription and updating the `max-revision-timeout-seconds` in your `KnativeServing` custom resource (CR). This applies the timeout changes across all Knative services, and you can deploy services with specific timeouts up to the maximum value set.

. In your subscription, set the value of `ROUTE_HAPROXY_TIMEOUT` to your required timeout in seconds:
+
.Example of setting the `ROUTE_HAPROXY_TIMEOUT` value to 900 seconds
[source,terminal]
----

> $ oc patch subscription serverless-operator -n openshift-serverless --type='merge' -p '{"spec": {"config": {"env": [{"name": "ROUTE_HAPROXY_TIMEOUT", "value": "900"}]}}}'

Reviewer 2: ```suggestion
$ oc patch subscription.operators.coreos.com serverless-operator -n openshift-serverless --type='merge' -p '{"spec": {"config": {"env": [{"name": "ROUTE_HAPROXY_TIMEOUT", "value": "900"}]}}}'
```

Reviewer 2: (with Serverless operator installed a "subscription" is ambiguous, as it can mean either `subscription.operators.coreos.com` , or `subscription.messaging.knative.dev` , so an API group is required here

====

. Set the maximum revision timeout as follows in the `KnativeServing` custom resource (CR):
+
.Example of `max-revision-timeout-seconds` set to 900 seconds
[source,yaml]
----
apiVersion: operator.knative.dev/v1beta1
kind: KnativeServing
metadata:
  name: knative-serving
spec:
  config:
    defaults:
      max-revision-timeout-seconds: "900"
----
+
Serverless operator automatically adjusts the activator's `terminationGracePeriod` to this value to avoid request termination in cases where activator pods are being terminated themselves.
+
You can verify that the timeout has been set by running the following command: 
+
[source,terminal]
----
$ oc get deployment activator -n knative-serving -o jsonpath="{.spec.template.spec.terminationGracePeriodSeconds}"
900
----

. Now you can deploy a Knative service with the desired timeouts less or equal to the maximum set:
+
.A service definition with timeouts set to 800 seconds
[source,yaml]
----
apiVersion: serving.knative.dev/v1
kind: Service
metadata:

>   name:

Reviewer 3: You don't need the `name` here but it could be helpful for the user.

spec:
  template:
    spec:
      timeoutSeconds: 800
      responseStartTimeoutSeconds: 800
----

> . Depending on your cloud provider, you may also need to set or adjust the load balancer timeout:

Reviewer 3: I would set this as an `Optional` step or reword it to something like `Adjust the load balancer timeout if necessary for your cloud provider by running the following command`.

+
.Example of load balancer timeout adjustment for AWS Classic LB 
[source,terminal]
----
$ oc -n openshift-ingress-operator patch ingresscontroller/default --type=merge --patch='{"spec":{"endpointPublishingStrategy": {"type":"LoadBalancerService", "loadBalancer": {"scope":"External", "providerParameters":{"type":"AWS", "aws": {"type":"Classic", "classicLoadBalancer": {"connectionIdleTimeout":"20m"}}}}}}}'

> ingresscontroller.operator.openshift.io/default patched

Reviewer 3: ```suggestion

```

Reviewer 3: So that the user can copy and paste the commands, I'd leave out the verification.

----

> [IMPORTANT]

Reviewer 3: I would move this up to the start of the procedure and provide more details. Does this adjustment need to be done at the start of the procedure or at this moment in the procedure? If you leave it here, you need a `+` on line 78 to indent it.

====
When using {SMProductShortName}, if the activator pod is stopped while a long-running request is in-flight, the request will be interrupted.
To avoid request interruptions, you must adjust the `terminationDrainDuration` value in the `ServiceMeshControlPlane`: 
[source,yaml]
----

> techPreview:

Reviewer 3: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#yaml-formatting-for-kubernetes-and-openshift-api-objects:~:text=Include%20the%20apiVersion%20and%20kind%20so%20that%20a%20user%20always%20knows%20the%20context%20of%20the%20YAML.

We need more fields for this YAML and #... to show that it is an excerpt.

.Procedure

. Set the value of `ROUTE_HAPROXY_TIMEOUT` in your subscription to your required timeout in seconds:
+
.Example of setting the `ROUTE_HAPROXY_TIMEOUT` value to 900 seconds
[source,terminal]
----
$ oc patch subscription.operators.coreos.com serverless-operator -n openshift-serverless --type='merge' -p '{"spec": {"config": {"env": [{"name": "ROUTE_HAPROXY_TIMEOUT", "value": "900"}]}}}'
----
+
`ROUTE_HAPROXY_TIMEOUT` is an enviroment managed by the Serverless Operator and by default is set to `600`. This subscription update causes pods to be redeployed in the `openshift-serverless` namespace.

[NOTE]
====
If you created your routes manually and disabled auto-generation with the annotation `serving.knative.openshift.io/disableRoute`, you can configure the timeouts directly in the route definitions.
====

. Set the maximum revision timeout as follows in the `KnativeServing` custom resource (CR):
+
.Example of `max-revision-timeout-seconds` set to 900 seconds
[source,yaml]
----
apiVersion: operator.knative.dev/v1beta1
kind: KnativeServing
metadata:
  name: knative-serving
spec:
  config:
    defaults:
      max-revision-timeout-seconds: "900"
----
+

> The Serverless Operator automatically adjusts the activator's `terminationGracePeriod` to the set maximum revision timeout value to avoid request termination in cases where activator pods are being terminated themselves.

Reviewer 4: ```suggestion
The Serverless Operator automatically adjusts the `terminationGracePeriod` value of the activator to the set maximum revision timeout value to avoid request termination in cases where activator pods are being terminated themselves.
```

https://www.ibm.com/docs/en/ibm-style?topic=grammar-possessives
> Do not use possessive ’s with inanimate objects.

+
You can verify that the timeout has been set by running the following command: 
+
[source,terminal]
----
$ oc get deployment activator -n knative-serving -o jsonpath="{.spec.template.spec.terminationGracePeriodSeconds}"
900
----

. Optional: Adjust the load balancer timeout if necessary for your cloud provider:
+
.Example of load balancer timeout adjustment for AWS Classic LB 
[source,terminal]
----
$ oc -n openshift-ingress-operator patch ingresscontroller/default --type=merge --patch='{"spec":{"endpointPublishingStrategy": {"type":"LoadBalancerService", "loadBalancer": {"scope":"External", "providerParameters":{"type":"AWS", "aws": {"type":"Classic", "classicLoadBalancer": {"connectionIdleTimeout":"20m"}}}}}}}'
----

> .  Deploy a Knative service with the desired timeouts less or equal to the maximum set:

Reviewer 4: ```suggestion
. Deploy a Knative service with the desired timeouts less or equal to the maximum set:
```

extra space

+

> .A service definition with timeouts set to 800 seconds

Reviewer 4: ```suggestion
.Example `Service` CR with timeouts set to 800 seconds
```

[source,yaml]
----
apiVersion: serving.knative.dev/v1
kind: Service
metadata:

>   name:

Reviewer 4: ```suggestion
  name: example-service-name
```

spec:
  template:
    spec:
      timeoutSeconds: 800
      responseStartTimeoutSeconds: 800
----
+
[IMPORTANT]
====

> When using {SMProductShortName}, if the activator pod is stopped while a long-running request is in-flight, the request will be interrupted.

Reviewer 4: ```suggestion
When using {SMProductShortName}, if the activator pod is stopped while a long-running request is in-flight, the request is interrupted.
```

> To avoid request interruptions, you must adjust the `terminationDrainDuration` value in the `ServiceMeshControlPlane`: 

Reviewer 4: ```suggestion
To avoid request interruptions, you must adjust the `terminationDrainDuration` value in the `ServiceMeshControlPlane` CR: 
```

- Add an object name. See, https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#object-references

[source,yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
#...
spec: 
  techPreview:
      meshConfig:
        defaultConfig:
          terminationDrainDuration: 1000s <1>
#... 
----
<1> Ensure that the value exceeds the request duration to avoid the Istio proxy shutdown, which would interrupt the request.
====

.Verification
If you are using Kourier, you can verify the current value of the timeout at the {ocp-product-title} route by running the following command:

[source,terminal]
----

> $ oc get route <route-name> -n knative-serving-ingress ess -o jsonpath="{.metadata.annotations.haproxy\.router\.openshift\.io/timeout}"

Reviewer 4: ```suggestion
$ oc get route <route_name> -n knative-serving-ingress ess -o jsonpath="{.metadata.annotations.haproxy\.router\.openshift\.io/timeout}"
```

User-replaced values need underscores.

+
[NOTE]
====
If you created your routes manually and disabled auto-generation with the `serving.knative.openshift.io/disableRoute` annotation, you can configure the timeouts directly in the route definitions.
====

. Set the maximum revision timeout in your `KnativeServing` CR:
+
.`KnativeServing` CR with `max-revision-timeout-seconds` set to 900 seconds
[source,yaml]
----
apiVersion: operator.knative.dev/v1beta1
kind: KnativeServing
metadata:
  name: knative-serving
spec:
  config:
    defaults:
      max-revision-timeout-seconds: "900"
#...
----
+
The Serverless Operator automatically adjusts the `terminationGracePeriod` value of the activator to the set maximum revision timeout value to avoid request termination in cases where activator pods are being terminated themselves.
+
. Optional: Verify that the timeout has been set by running the following command: 
+
[source,terminal]
----
$ oc get deployment activator -n knative-serving -o jsonpath="{.spec.template.spec.terminationGracePeriodSeconds}"
----

> . If necessary for your cloud provider, adjust the load balancer timeout:

Reviewer: ```suggestion
. If necessary for your cloud provider, adjust the load balancer timeout by running the following command:
```

====
If you created your routes manually and disabled auto-generation with the `serving.knative.openshift.io/disableRoute` annotation, you can configure the timeouts directly in the route definitions.
====

. Set the maximum revision timeout in your `KnativeServing` CR:
+
.`KnativeServing` CR with `max-revision-timeout-seconds` set to 900 seconds
[source,yaml]
----
apiVersion: operator.knative.dev/v1beta1
kind: KnativeServing
metadata:
  name: knative-serving
spec:
  config:
    defaults:
      max-revision-timeout-seconds: "900"
#...
----
+
The Serverless Operator automatically adjusts the `terminationGracePeriod` value of the activator to the set maximum revision timeout value to avoid request termination in cases where activator pods are being terminated themselves.
+
. Optional: Verify that the timeout has been set by running the following command: 
+
[source,terminal]
----
$ oc get deployment activator -n knative-serving -o jsonpath="{.spec.template.spec.terminationGracePeriodSeconds}"
----

. If necessary for your cloud provider, adjust the load balancer timeout by running the following command:
+
.Load balancer timeout adjustment for AWS Classic LB 
[source,terminal]
----

> $ oc -n openshift-ingress-operator patch ingresscontroller/default --type=merge --patch='

Reviewer 5: Hi, these need to be broken apart with a `\` as described in the [guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#backslash-for-long-lines-of-code). That ensures that they display syntax highlighting properly and are copied correctly if someone uses the "click to copy" function. For example:

```
$ oc -n openshift-ingress-operator patch ingresscontroller/default --type=merge --patch= \
  '{"spec":{"endpointPublishingStrategy":  \
  {"scope":"External", "providerParameters":{"type":"AWS", "aws":  \
  {"type":"Classic", "classicLoadBalancer":  \
  {"connectionIdleTimeout":"20m"}}}}}}}'
```
